### DISCUSSION TITLE
NTLM Authentication failed when build with Openssl

### DISCUSSION THREAD

**Discussion by fengyue520** [curl/curl#9579] [Q&A] [answered]
Hello everyone, I found that when using the configuration of "Lib Release" to compile curl, using curl.exe to access an Exchange Server with EP(Extended Protection) enabled and set to require, the following error will occur.

```
*   Trying IP..
* Connected to exchange2019.mydomain (ip) port 443 (#0)
* ALPN, offering http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
* ...
* Server auth using NTLM with user 'user'
> GET /EWS/Exchange.asmx HTTP/1.1
> Host: exchange2019.mydomain
> Authorization: NTLM TlRMTVNTUAABAAAAB4IIogAAAAAAAAAAAAAAAAAAAAAKAGFKAAAADw==
> User-Agent: curl/7.82.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Server: Microsoft-IIS/10.0
< request-id: -
< WWW-Authenticate: NTLM TlRMTVNTUAACAAAAGgAaADgAAAAFgomird/a2+RnudMAAAAAAAAAAAwBDAFSAAAACgBjRQAAAA9FAFgAQwBIAEEATgBHAEUAMgAwADEAOQAwAAIAGgBFAFgAQwBIAEEATgBHAEUAMgAwADEAOQAwAAEAHgBXAEkATgAtAEMAVgBFAFYASABJAFIAQQBDADEANgAEADAAZQB4AGMAaABhAG4AZwBlADIAMAAxADkALgBzAHUAegBoAG8AdQAuAHoAbwBvAG0AAwBQAFcASQBOAC0AQwBWAEUAVgBIAEkAUgBBAEMAMQA2AC4AZQB4AGMAaABhAG4AZwBlADIAMAAxADkALgBzAHUAegBoAG8AdQAuAHoAbwBvAG0ABQAwAGUAeABjAGgAYQBuAGcAZQAyADAAMQA5AC4AcwB1AHoAaABvAHUALgB6AG8AbwBtAAcACAA00STKa8/YAQAAAAA=
< X-OWA-Version: 15.2.1118.12
< WWW-Authenticate: Negotiate
< X-Powered-By: ASP.NET
< X-FEServer: -
< Date: Fri, 23 Sep 2022 16:44:45 GMT
< Content-Length: 0
<
* Connection #0 to host exchange2019.mydomain left intact
* Issue another request to this URL: 'https://exchange2019.mydomain/EWS/Exchange.asmx'
* Found bundle for host exchange2019.mydomain: 0xde2be0 [serially]
* Re-using existing connection! (#0) with host exchange2019.mydomain
* Connected to exchange2019.mydomain (ip) port 443 (#0)
* Server auth using NTLM with user 'del.service@exchange2019.suzhou'
> GET /EWS/Exchange.asmx HTTP/1.1
> Host: exchange2019.mydomain
> Authorization: NTLM TlRMTVNTUAADAAAAGAAYALIAAADGAcYBygAAAAAAAABYAAAASABIAFgAAAASABIAoAAAAAAAAACQAgAABYKIogoAYUoAAAAPgsIfFRuxe85ScnUFSKBRbWQAZQBsAC4AcwBlAHIAdgBpAGMAZQBAAGUAeABjAGgAYQBuAGcAZQAyADAAMQA5AC4AcwB1AHoAaABvAHUALgB6AG8AbwBtAEgAVwAwADAAMAA2ADkAMwA4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAK1L4vSnd5H96f45yjTd6J8BAQAAAAAAADTRJMprz9gBkrW/x2+BMh4AAAAAAgAaAEUAWABDAEgAQQBOAEcARQAyADAAMQA5ADAAAQAeAFcASQBOAC0AQwBWAEUAVgBIAEkAUgBBAEMAMQA2AAQAMABlAHgAYwBoAGEAbgBnAGUAMgAwADEAOQAuAHMAdQB6AGgAbwB1AC4AegBvAG8AbQADAFAAVwBJAE4ALQBDAFYARQBWAEgASQBSAEEAQwAxADYALgBlAHgAYwBoAGEAbgBnAGUAMgAwADEAOQAuAHMAdQB6AGgAbwB1AC4AegBvAG8AbQAFADAAZQB4AGMAaABhAG4AZwBlADIAMAAxADkALgBzAHUAegBoAG8AdQAuAHoAbwBvAG0ABwAIADTRJMprz9gBBgAEAAIAAAAIADAAMAAAAAAAAAABAAAAACAAABgLA+5SZz3i8WR/gb0t6voJxKo9AMFMmdNGufRJq3ETCgAQAAAAAAAAAAAAAAAAAAAAAAAJADYASABUAFQAUAAvAGUAeABjAGgAYQBuAGcAZQAyADAAMQA5AC4AegBpAHAAbwB3AC4AYwBvAG0AAAAAAAAAAAA=
> User-Agent: curl/7.82.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Server: Microsoft-IIS/10.0
< request-id: -
< X-OWA-Version: 15.2.1118.12
< WWW-Authenticate: Negotiate
* NTLM handshake rejected
* Authentication problem. Ignoring this.
< WWW-Authenticate: NTLM
< X-Powered-By: ASP.NET
< X-FEServer: WIN-CVEVHIRAC16
< Date: Fri, 23 Sep 2022 16:44:48 GMT
< Content-Length: 0
<
* Connection #0 to host exchange2019.mydomain left intact
```

error: NTLM handshake rejected

I found that the previous issue raised the same problem and it has been solved https://github.com/curl/curl/issues/3280. I read the source code and found that this place handles the logic related to channel binding, but when build with openssl, this logic does not come in. It just used for Schannel.

![image](https://user-images.githubusercontent.com/48266771/191925633-bff7f53e-42d1-4788-bb31-5d30eb05e447.png)

At present, I'm not entirely sure if using Openssl as the ssl backend can't support EP or if this is a bug. Can anyone provide some advice, is it possible to use Openssl to support EP?

Much thanks.

**Comment by bagder** [comment]
I'm not sure anyone here knows anything about this "Extended Protection" you speak of, so we cannot really tell what it requires and why this curl build doesn't work with it.

**Reply by fengyue520** [reply]
sorry， "[Extended Protection](https://learn.microsoft.com/en-us/dotnet/framework/wcf/feature-details/extended-protection-for-authentication-overview)" is a windows unique configuration, here is the overview: https://learn.microsoft.com/en-us/iis/configuration/system.webserver/security/authentication/windowsauthentication/extendedprotection/

**Reply by bagder** [reply]
And yet I bet the answer to this issue is not in the "overview" but deep inside the specific details of how it works.

**Reply by fengyue520** [reply]
Yes, I guess so. And I tested it with the linux version of libcurl, the problem is the same.
I have briefly researched the EP before, and it seems that the API of the windows system needs to be used, but I have not researched it thoroughly, and there is no conclusion that whether openssl can be used to support this function.
So I'm here to ask if anyone who's looked into this feature can offer some valuable advice.

**Comment by fengyue520** [comment, marked as answer]
Sorry for not replying for a long time, I have found a solution to this problem.

Let me talk about the root cause of the problem first. When compiling libcurl with OpenSSL & SSPI, if the exchange server enables EPA ([Extend Protected Auth](https://learn.microsoft.com/en-us/dotnet/framework/wcf/feature-details/extended-protection-for-authentication-overview)), the NTLM handshake process will fail because there is no [CBT](https://learn.microsoft.com/en-us/archive/blogs/openspecification/ntlm-and-channel-binding-hash-aka-extended-protection-for-authentication) field.

But if you use Schannel & SSPI to compile libcurl, there is no such problem, because this problem has been solved in this issue https://github.com/curl/curl/issues/3280.

So my solution is very simple. Since Schannel calculates CBT internally, but OpenSSL does not, I calculate CBT directly outside and bring this field in the NTLM handshake. The calculation method of CBT refers to this open source project https://github.com/eurogiciel-oss/freerdp/tree/ff1454a973fa1f5a938dc4af0563e89e95c3489f. After testing, this change can solve the problem.

I put the changes below, it can give some reference for people who encounter the same problem.
![image](https://user-images.githubusercontent.com/48266771/205504656-5270d197-a24b-4358-a772-94cb984cda42.png)
![image](https://user-images.githubusercontent.com/48266771/205504673-df4ed1c3-1f18-4d01-8993-d6755dba733a.png)
![image](https://user-images.githubusercontent.com/48266771/205504692-ee46df69-a275-4b8f-b65c-6a2535b9eca1.png)
